### PR TITLE
Serialization refactoring.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi @mzhurovich!

As a side effect of refactoring serialization, I managed to speed up `ParseJSON` by replacing `std::string path` by a vector of `KindaVariant<const char*, int>`.

```
ubuntu@hetz:~/clean$ for branch in master serialization ; do \
>   git clone -b $branch https://github.com/dkorolev/Current $branch ; \
>   (cd $branch/examples/Benchmark/generic; \
>    mkdir -p .current ; \
>    clang++-3.5 -std=c++11 -O3 -DNDEBUG -o .current/run run.cc -pthread && \
>    for i in $(seq 3) ; do ./.current/run --seconds=20 --scenario=json --json=parse ; done); \
> done
Cloning into 'master'...
4828.15 QPS.
4824.3 QPS.
4820.1 QPS.
Cloning into 'serialization'...
5548.05 QPS.
5574.8 QPS.
5565.25 QPS.
```

That's 15% improvement on the simple benchmark; the improvement on a real 3+GB file is about 30%.

Next steps call for a sync-up. I'll probably last two hours, or my morning your evening.

Thanks!
Dima

### CODE REVIEW

File: TypeSystem/Serialization/json/variant.h [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+class VB {
+ public:
+  virtual ~VB() = default;
+  virtual void Deserialize(JSONParser<JSON_FORMAT>& json_parser,
+                           IHasUncheckedMoveFromUniquePtr& destination) = 0;
+};
+
+// `VC1`: variant case implementation #1, generic.
+template <class JSON_FORMAT, typename T>
+class VC1 : public VB<JSON_FORMAT> {
```
Comment (mzhurovich) [inline TypeSystem/Serialization/json/variant.h]: Do you think it's worth to use ultra-short class names here?

File: TypeSystem/Serialization/serialization.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+
+#ifndef CURRENT_TYPE_SYSTEM_SERIALIZATION_SERIALIZATION_H
+#define CURRENT_TYPE_SYSTEM_SERIALIZATION_SERIALIZATION_H
+
+#include <iostream>  // TODO(dkorolev): DIMA FIXME! Remove as we're refactoring our `binary` format.
```
Comment (mzhurovich) [inline TypeSystem/Serialization/serialization.h]: Do you still need it here, as you killed all the binary (de)serialization code?

File: TypeSystem/Serialization/json/json.h [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+struct JSONValueAssignerImpl {
+  static void AssignValue(rapidjson::Value& destination, current::copy_free<T> value) { destination = value; }
+};
+
+template <class JSON_FORMAT>
+class JSONStringifier final {
+ public:
+  JSONStringifier() {
+    // Can't assign in the initializer list, `current_` is the field before `document_`.
+    // And I've confirmed keeping `document_` first results in performance degradation. -- D.K.
```
Comment (mzhurovich) [inline TypeSystem/Serialization/json/json.h]: Interesting. Do you know, why?

File: TypeSystem/Serialization/json/variant.h (hunk 1 above)
Comment (dkorolev) [inline TypeSystem/Serialization/json/variant.h]: :+1: bcbebe7a80db9397c710f74863a1f9aeae1a54c9

File: TypeSystem/Serialization/serialization.h (hunk 2 above)
Comment (dkorolev) [inline TypeSystem/Serialization/serialization.h]: :+1: 4a0503d6cf032cefbb6971472af5263519ad7a6d

File: TypeSystem/Serialization/json/json.h (hunk 3 above)
Comment (dkorolev) [inline TypeSystem/Serialization/json/json.h]: Alignment, I'd assume.

**Comment by dkorolev** [comment]
@mzhurovich PTAL -- addressed your comments and renamed `PatchJSON` into `PatchObjectWithJSON`.

**Comment by mzhurovich** [comment]
@dkorolev 👍

**Pull request merged**